examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts

```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
         }
 
         const doTheMigration = async () => {
-            // It's possible that our modelLoader is older and doesn't understand the new acceptedVersion.  Currently
-            // this fails the migration gracefully and emits an event so the app developer can know they're stuck.
-            // Ideally the app developer would find a way to acquire a new ModelLoader and move forward, or at least
-            // advise the end user to refresh the page or something.
-            // TODO: Does the app developer have everything they need to dispose gracefully when recovering with a new
-            // ModelLoader?
-            const migrationSupported = await this.modelLoader.supportsVersion(acceptedVersion);
-            if (!migrationSupported) {
-                this.emit("migrationNotSupported", acceptedVersion);
-                this._migrationP = undefined;
-                return;
-            }
+            // doTheMigration() is called at the start of migration and should only resolve in two cases. First, is if
+            // we complete the migration. Second, is if we disconnect during the migration process. In both cases we
+            // should re-enter the state machine and take the appropriate action (see then() block below).
 
-            const createResponse = await this.modelLoader.createDetached(acceptedVersion);
-            const migratedModel: IMigratableModel = createResponse.model;
-
-            const exportedData = await migratable.exportData();
-
-            // TODO: Is there a reasonable way to validate at proposal time whether we'll be able to get the exported
-            // data into a format that the new model can import?  If we can determine it early, then clients with old
-            // ModelLoaders can use that opportunity to dispose early and try to get new ModelLoaders.
-            let transformedData: unknown;
-            if (migratedModel.supportsDataFormat(exportedData)) {
-                // If the migrated model already supports the data format, go ahead with the migration.
-                transformedData = exportedData;
-            } else if (this.dataTransformationCallback !== undefined) {
-                // Otherwise, try using the dataTransformationCallback if provided to get the exported data into
-                // a format that we can import.
-                try {
-                    transformedData = await this.dataTransformationCallback(exportedData, migratedModel.version);
-                } catch {
-                    // TODO: This implies that the contract is to throw if the data can't be transformed, which isn't
-                    // great.  How should the dataTransformationCallback indicate failure?
+            const prepareTheMigration = async () => {
+                if (this._preparedDetachedModel !== undefined) {
```

Comment:
Who ensures that every client will land on the same detached model, without any data loss? Is it modelLoader? I'm curious what are some of the contraints there?

Reply:
That is what `completeMigration()` takes care of. That function will ensure a single client (through TaskManager) attaches their local detached model and then shares it with the other clients

Reply:
As I understood completeMigration completes the migration of the detached mode? My question was mostly who ensures that "detached" model

> Some general thoughts:
> 
> 1. It would be nice if we had "migration" package that essentially hides all this complexity of variations and unhappy paths that can play out with clear success/fail outcomes.
> 2. It would be nice to understand use cases out migration should support. Then it may be easier to reason around success/fail outcomes. For example if we have use case scenario where user wants migration to occur only if certian condition was met (say all clients can upgrade), we could reason what building blocks we need in place.
> 3. From perspective of TaskManager it would be nice to abstract some helper functions that know how to handle various error cases as they execute async block.

Btw these are general comments going forward, I understand that scope of this PR is mostly focused on TaskManager for migration.